File: ChatGPT_BP/20_1a_BP.py
import re
from urllib.parse import urlparse, urljoin
from flask import Flask, request, redirect, abort

app = Flask(__name__)

# Compiled regular expression for trusted domains
TRUSTED_DOMAIN_REGEX = re.compile(r'^(https?://)?(www\.)?(trusted\.com|anothertrusted\.com)($|/)')

# Default redirect URL
DEFAULT_REDIRECT_URL = 'https://default.com'

def is_trusted_domain(url):
    """
    Check if the given URL matches the trusted domain regular expression.
    """
    return TRUSTED_DOMAIN_REGEX.match(url) is not None

def sanitize_url(url):
    """
    Sanitize the URL to prevent open redirects.
    """
    parsed_url = urlparse(url)
    return urljoin(request.host_url, parsed_url.path)

@app.route('/redirect', methods=['GET'])
def handle_redirect():
    """
    Handle URL redirection based on the target URL.
    """
    target_url = request.args.get('url')
    
    if not target_url:
        abort(400, "No URL provided")

    # Check if the target URL matches the trusted domain
    if is_trusted_domain(target_url):
        # Sanitize the URL before redirecting
        safe_url = sanitize_url(target_url)
        return redirect(safe_url)
    else:
        # Redirect to the default URL if the target URL does not match
        return redirect(DEFAULT_REDIRECT_URL)

if __name__ == '__main__':
    app.run(debug=True)

File: ChatGPT_BP_Java/CWE-20/20_1a_BP.java
import java.util.regex.Pattern;
import java.util.regex.Matcher;
import java.net.URL;
import java.net.URISyntaxException;

public class URLRedirector {

    // Compile the regular expression once when the module is loaded
    private static final Pattern TARGET_URL_PATTERN = Pattern.compile("^https://(www\\.)?trusteddomain\\.com/.*");
    private static final String DEFAULT_REDIRECT_URL = "https://www.defaultdomain.com";
    private static final String[] TRUSTED_DOMAINS = { "trusteddomain.com", "anothertrusted.com" };

    // Method to sanitize the URL
    public static String sanitizeUrl(String url) {
        try {
            URL validUrl = new URL(url);
            validUrl.toURI(); // Ensure it's a valid URI
            return validUrl.toString();
        } catch (URISyntaxException | java.net.MalformedURLException e) {
            // Invalid URL: fallback to default URL
            return DEFAULT_REDIRECT_URL;
        }
    }

    // Method to check if the URL is from a trusted domain
    public static boolean isTrustedDomain(String url) {
        try {
            URL validUrl = new URL(url);
            String host = validUrl.getHost();
            for (String trustedDomain : TRUSTED_DOMAINS) {
                if (host.endsWith(trustedDomain)) {
                    return true;
                }
            }
        } catch (java.net.MalformedURLException e) {
            // Malformed URL
        }
        return false;
    }

    // Method to perform the redirect based on the target URL
    public static String handleRedirect(String targetUrl) {
        targetUrl = sanitizeUrl(targetUrl); // Sanitize input URL

        // Check if target URL matches the regular expression pattern
        Matcher matcher = TARGET_URL_PATTERN.matcher(targetUrl);
        if (matcher.matches() && isTrustedDomain(targetUrl)) {
            return targetUrl; // Use the target URL if it matches the pattern
        } else {
            return DEFAULT_REDIRECT_URL; // Use default URL if not matching
        }
    }

    public static void main(String[] args) {
        // Example target URL
        String targetUrl = "https://www.trusteddomain.com/path/to/resource";
        String redirectUrl = handleRedirect(targetUrl);

        System.out.println("Redirecting to: " + redirectUrl);

        // Example of an untrusted URL
        targetUrl = "https://www.untrusted.com/unsafe/path";
        redirectUrl = handleRedirect(targetUrl);
        System.out.println("Redirecting to: " + redirectUrl);
    }
}
